File: app/src/androidTest/java/com/github/fo2rist/mclaren/MainActivityTest.java
package com.github.fo2rist.mclaren;

import android.support.test.espresso.Espresso;
import android.support.test.runner.AndroidJUnit4;

import com.github.fo2rist.mclaren.pages.NewsfeedPage;

import org.junit.Test;
import org.junit.runner.RunWith;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static com.github.fo2rist.mclaren.utilities.CustomViewAssertions.displayed;
import static org.hamcrest.Matchers.not;

@RunWith(AndroidJUnit4.class)
public class MainActivityTest extends BaseMainActivityTest {

    @Test
    public void testMainScreenLayout() throws Exception {
        mainPage.onContentFrame()
                .check(displayed());
        mainPage.onToolbar()
                .check(displayed());
        mainPage.onFloatingButton()
                .check(displayed());
    }

    @Test
    public void testSideMenu() throws Exception {
        mainPage.onMenuNewsfeed()
                .check(matches(not(isDisplayed())));

        mainPage.openNavigationDrawer();

        mainPage.onMenuNewsfeed()
                .check(displayed());
        mainPage.onMenuCircuits()
                .check(displayed());
        mainPage.onMenuDrivers()
                .check(displayed());
        mainPage.onMenuCar()
                .check(displayed());
    }

    @Test
    public void testOptionsMenu() throws Exception {
        Espresso.openActionBarOverflowOrOptionsMenu(context);
        onView(withText("Settings"))
                .check(displayed());
    }

    @Test
    public void testNavigationToSamePage() throws Exception {
        NewsfeedPage newsfeedPage = new NewsfeedPage();
        newsfeedPage.onNewsList()
                .check(displayed());
        mainPage.navigateToMenuItem(R.id.nav_newsfeed);
        newsfeedPage.onNewsList()
                .check(displayed());
    }

}

File: app/src/androidTest/java/com/github/fo2rist/mclaren/tests/MainActivityTest.kt
package com.github.fo2rist.mclaren.tests

import android.net.Uri
import androidx.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu
import androidx.test.espresso.intent.Intents.intended
import androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent
import androidx.test.espresso.intent.matcher.IntentMatchers.hasData
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.fo2rist.mclaren.R
import com.github.fo2rist.mclaren.pages.FeedPage
import com.github.fo2rist.mclaren.ui.circuitsscreen.CircuitDetailsActivity
import com.github.fo2rist.mclaren.ui.previewscreen.PreviewActivity
import com.github.fo2rist.mclaren.ui.transmissionscreen.TransmissionActivity
import org.hamcrest.Matchers.allOf
import org.hamcrest.Matchers.any
import org.joda.time.DateTime
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class MainActivityTest : BaseMainActivityTest() {

    @Test
    fun testSideMenu() {
        mainPage {
            menuStories { isNotDisplayed() }

            openNavigationDrawer()

            menuStories { isDisplayed() }
            menuTeamTwitter { isDisplayed() }
            menuCircuits { isDisplayed() }
            menuDrivers { isDisplayed() }
            menuCar { isDisplayed() }
            menuOfficialSite { isDisplayed() }
        }
    }

    @Test
    fun testFloatingButtonUpcomingEvent() {
        val upcomingEventName = "TEST UPCOMING GP"
        val upcomingEventTime = DateTime().plusMinutes(61)
        interceptIntents()
        activity.runOnUiThread { activity.showUpcomingEventButton(upcomingEventName, upcomingEventTime) }

        mainPage {
            upcomingEventButton {
                isDisplayed()
                matches {
                    containsText(upcomingEventName)
                    containsText("1H")
                }
                perform { click() }
            }
        }
        intended(hasComponent(CircuitDetailsActivity::class.java.name))
    }

    @Test
    fun testFloatingButtonTransmission() {
        interceptIntents()
        activity.runOnUiThread { activity.showTransmissionButton() }
        mainPage {
            transmissionButton {
                isDisplayed()
                perform { click() }
            }
        }
        intended(hasComponent(TransmissionActivity::class.java.name))
    }

    @Test
    fun testOptionsMenu() {
        interceptIntents()

        mainPage {
            openActionBarOverflowOrOptionsMenu(context)
            optionMenuAbout.click()
        }

        intended(hasComponent(PreviewActivity::class.java.name))
    }

    @Test
    fun testNavigationToSamePage() {
        val feedPage = FeedPage()

        feedPage.feedList.isDisplayed()

        mainPage.navigateToMenuItem(R.id.nav_stories)

        feedPage.feedList.isDisplayed()
    }

    @Test
    fun testFooterMenuActionCar() {
        interceptIntents()

        mainPage.navigateToFooterMenuItem(R.id.nav_car)

        intended(hasData(any(Uri::class.java)))
    }

    @Test
    fun testFooterMenuActionOfficialSite() {
        interceptIntents()

        mainPage.navigateToFooterMenuItem(R.id.nav_official_site)

        intended(hasData(any(Uri::class.java)))
    }
}
